tests/ga/test_extension.py

```diff
@@ -1464,24 +1464,45 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
         self._assert_no_handler_status(protocol.report_vm_status)
 
     @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_update_command')
-    def test_upgrade_failure(self, patch_get_update_command, *args):
+    def test_extension_upgrade_failure_when_new_version_update_fails(self, patch_get_update_command, *args):
         """
-        Extension upgrade failure should not be retried
+        When the update command of the new extension fails, it should result in the new extension removed and the
+        old extension disabled. On the next goal state, the entire upgrade scenario should be retried (once),
```

Comment:
Quick question, if the update fails then the old one is disabled, on a new GS would disable for the old one be called again?
Previously because of the Failed state we were not calling Disable again but calling the enable on the new one directly

Reply:
Yes, disable would be called on the old extension again because this time we'd go into [this code path](https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/ga/exthandlers.py#L495).

Reply:
I wonder if this would throw extensions in a weird state because they may or may not be handling a scenario where disable is called multiple times. We'll have to make sure they know about this change so that they can modify their code too if needed

Reply:
The contract with extension publishers is that each operation should be idempotent. From the document:

> Every time the machine restarts (which in turn will restart the Azure Agent) or a new goal state is received, the Azure Agent will try to set the machine to the latest goal state. Thus it might invoke the enabled\disabled commands multiple times even if the handler is already enabled\disabled. So the enable and disable commands need to be idempotent

Reply:
I'd say we can go ahead, since disable is supposed to be idempotent. From our publisher's doc:
> Unlike the install state, the Azure Agent will not maintain the enabled\disabled state of the handler. Every time the machine restarts (which in turn will restart the Azure Agent) or a new goal state is received, the Azure Agent will try to set the machine to the latest goal state. Thus it might invoke the enabled\disabled commands multiple times even if the handler is already enabled\disabled. So the enable and disable commands need to be idempotent i.e. if the handler is already enabled and the enable command is invoked again, the command should check if all the processes are running as expected, if yes, then the command should just exit with a success code.

Reply:
Nice! I didnt know about disable, I thought we just said that enable has to be idempotent. THansk for clarifying it guys :)